tests/bsim/bluetooth/host/adv/extended/src/ext_adv_advertiser.c

```diff
@@ -0,0 +1,229 @@
+/**
+ * Copyright (c) 2024 Croxel, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/kernel.h>
+
+#include "common.h"
+
+#include "bs_types.h"
+#include "bs_tracing.h"
+#include "time_machine.h"
+#include "bstests.h"
+
+#include <zephyr/types.h>
+#include <zephyr/sys/printk.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+
+extern enum bst_result_t bst_result;
+
+static struct bt_conn *g_conn;
+
+CREATE_FLAG(flag_connected);
+CREATE_FLAG(flag_conn_recycled);
+
+static void common_init(void)
+{
+	int err;
+
+	err = bt_enable(NULL);
+
+	if (err) {
+		FAIL("Bluetooth init failed: %d\n", err);
+		return;
+	}
+	printk("Bluetooth initialized\n");
+}
+
+static void create_ext_adv_set(struct bt_le_ext_adv **adv, bool connectable)
+{
+	int err;
+
+	printk("Creating extended advertising set...");
+
+	const struct bt_le_adv_param *adv_param = connectable ?
+		BT_LE_EXT_ADV_CONN_NAME : BT_LE_EXT_ADV_NCONN_NAME;
+
+	err = bt_le_ext_adv_create(adv_param, NULL, adv);
+	if (err) {
+		printk("Failed to create advertising set: %d\n", err);
+		return;
+	}
+	printk("done.\n");
+}
+
+static void start_ext_adv_set(struct bt_le_ext_adv *adv)
+{
+	int err;
+
+	printk("Starting Extended Advertising...");
+	err = bt_le_ext_adv_start(adv, BT_LE_EXT_ADV_START_DEFAULT);
+	if (err) {
+		printk("Failed to start extended advertising: %d\n", err);
+		return;
+	}
+	printk("done.\n");
+}
+
+static void stop_ext_adv_set(struct bt_le_ext_adv *adv)
+{
+	int err;
+
+	printk("Stopping Extended Advertising...");
+	err = bt_le_ext_adv_stop(adv);
+	if (err) {
+		printk("Failed to stop extended advertising: %d\n",
+		       err);
+		return;
+	}
+	printk("done.\n");
+}
+
+static void delete_adv_set(struct bt_le_ext_adv *adv)
+{
+	int err;
+
+	printk("Delete extended advertising set...");
+	err = bt_le_ext_adv_delete(adv);
+	if (err) {
+		printk("Failed Delete extended advertising set: %d\n", err);
+		return;
+	}
+	printk("done.\n");
+}
+
+static void disconnect_from_target(void)
+{
+	int err;
+
+	printk("Disconnecting...\n");
+
+	err = bt_conn_disconnect(g_conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+	if (err) {
+		FAIL("BT Disconnect failed: %d\n", err);
+		return;
+	}
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err != BT_HCI_ERR_SUCCESS) {
+		FAIL("Failed to connect to %s: %u\n", addr, err);
+		return;
+	}
+
+	printk("Connected to %s\n", addr);
+	g_conn = bt_conn_ref(conn);
```

Comment:
I think `g_conn` might already be holding a reference. The reference comes from `bt_conn_le_create`. Can you check here that `g_conn` is NULL?

Might you be interested the Bluetooth testlib?

We already have [bt_testlib_conn_unref](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/include/testlib/conn.h#L73), which is an conn_unref with atomics and safety checks. I've been thinking that we should add `bt_testlib_conn_move(**dst_connp, **src_connp)` and `bt_testlib_conn_clone(**dst_connp, *conn)`.

Using and extending testlib is just a suggestion.

Reply:
Thanks for the heads-up. I'll check Bluetooth testlib and integrate it into this sample if practical

Reply:
Small follow-up: `g_conn` here is static to this file. Considering this file is the advertiser, how would `bt_conn_le_create()` influence `g_conn` here?

Reply:
You are right. I read across file when I saw `bt_conn_le_create`. It's a bit unfortunate that `bt_le_ext_adv_start` API is not consistent with `bt_conn_le_create` and does not return a connection object. `g_conn` will probably be null here.

Reply:
No worries - I'll add the sanity check to better communicate the user that bt_conn references should not be overriden.